apostrophecms/apostrophe
# OpenTelemetry

## Summary

**Integrate OpenTelemetry instrumentation in Apostrophe** 

The main goal is to make Apostrophe "telemetry ready" without or with the little possible overhead while not used.

**Features:**
- Boot tracing
- Per request tracing
- Tasks tracing
- Key points instrumentation with a high effort to add standard and Apostrophe owned attributes, allowing trace backend filtering
- Relationships (graph, based on active context) to allow operations (including filtered) to be viewed in the appropriate context
- Allow application level instrumentation

**Integration details:**

Single dependency point with OpenTelemetry API - `apos.telemetry` pseudo module. Direct dependency with OpenTelemetry Semantic Attributes per module. `apos.telemetry` exposes a simplified API (OTel API wrappers).

**Backward compatibility**

The integration is fully backward compatible with the previous platform functionality and exposed public APIs. 
The only change that might be considered non BC is `@apostrophecms/task:usage() signature. However this shouldn't be considered a public API and affects only the internal development so it can't be declared as breaking change.
Another notable (BC) change is the `run` emitter passing down the application Root span the event listeners.

**Operation prefixes:**

In order to keep the visual inspection experience high, the following prefixes for operation names were introduced (additionally to the naming convention standards):

- `aspostrophe:` boot sequence operations - `apostrophe:boot`, `apostrophe:boot:primary`, `apostrophe:boot:worker`
- `event:` event emitter operations
- `handler:` event listeners operations
- `db:` query related operations
- `render:` template render related operations
- `task:` task related operations

Closes PRO-2660, PRO-2664, PRO-2665, PRO-2666, PRO-2661

## What are the specific steps to test this change?

- Introduce OpenTelemetry SDK dependencies and configure the application. Pass down a shutdown handler as `beforeExit` configuration.
- Install collector backend (e.g. Jaeger).
- Boot up the application, login, manage, navigate. Execute Apostrophe tasks.
- Check the collector backend. 

## What kind of change does this PR introduce?
*(Check at least one)*


## Make sure the PR fulfills these requirements:
